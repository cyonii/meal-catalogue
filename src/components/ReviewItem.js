import { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import ListGroup from 'react-bootstrap/ListGroup';
import Button from 'react-bootstrap/Button';

const ReviewItem = ({ review }) => {
  const { content } = review;
  const sliceLength = 250;
  const [showAll, setShowAll] = useState(false);
  const showableContent = showAll ? content : content.slice(0, sliceLength);
  const splittedContent = showableContent.split('\n');
  const listItem = useRef(null);

  const toggleCollapse = () => {
    setShowAll(!showAll);
    listItem.current.scrollIntoView(true, { behavior: 'smooth' });
  };

  return (
    <ListGroup.Item as="li" ref={listItem} className="px-0 bg-transparent text-white border-0 border-bottom border-secondary" key={review.id} data-testid="review-item">
      <h6 className="text-info mb-0 lh-1">{`${review.author}`}</h6>
      <div className="d-flex justify-content-between align-items-center mb-3">
        <small className="text-muted">{`Rating: ${review.author_details.rating}`}</small>
        {showAll && (
        <Button variant="link" className="text-light lh-1 py-0 pb-1 text-decoration-none" onClick={toggleCollapse}>
          <small>{showAll ? 'Show less' : 'Show more'}</small>
        </Button>
        )}
      </div>

      {splittedContent.map((line, index) => (
        <p className="mb-2 opacity-75 review-text" key={Math.random()}>
          {line}
          { (!showAll && index === splittedContent.length - 1 && content.length > sliceLength) && <span className="text-muted fw-bold">...</span> }
        </p>
      ))}
      { content.length > sliceLength
      && (
      <div className="d-flex mt-1 opacity-100">
        <Button variant={showAll ? 'light' : 'info'} className="lh-1 py-0 pb-1 rounded-0 text-decoration-none" onClick={toggleCollapse}>
          <small className="fw-bold" style={{ fontSize: '13px' }}>
            {showAll ? 'Show less' : 'Show more'}
          </small>
        </Button>
      </div>
      ) }
    </ListGroup.Item>
  );
};

ReviewItem.propTypes = {
  review: PropTypes.instanceOf(Object).isRequired,
};

export default ReviewItem;
