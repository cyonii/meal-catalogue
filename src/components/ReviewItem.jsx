import { useState, useRef } from "react";
import PropTypes from "prop-types";

const ReviewItem = ({ review }) => {
  const { content } = review;
  const sliceLength = 250;
  const [showAll, setShowAll] = useState(false);
  const showableContent = showAll ? content : content.slice(0, sliceLength);
  const splittedContent = showableContent.split("\n");
  const listItem = useRef(null);

  const toggleCollapse = () => {
    setShowAll(!showAll);
    listItem.current.scrollIntoView(true, { behavior: "smooth" });
  };

  return (
    <li
      ref={listItem}
      className="py-3 text-slate-50 border-b border-slate-900"
      key={review.id}
      data-testid="review-item"
    >
      <h6 className="text-cyan-500 mb-0 leading-none">{`${review.author}`}</h6>
      <div className="flex justify-between items-center mb-3">
        <small>
          <span className="text-slate-500">Rating </span>
          {`${review.author_details.rating}`}
        </small>
        {showAll && (
          <button
            className="text-slate-300 leading-none py-0 pb-1"
            onClick={toggleCollapse}
          >
            <small>{showAll ? "Show less" : "Show more"}</small>
          </button>
        )}
      </div>

      {splittedContent.map((line, index) => (
        <p className="mb-2 opacity-75" key={Math.random()}>
          {line}
          {!showAll &&
            index === splittedContent.length - 1 &&
            content.length > sliceLength && (
              <span className="text-slate-500 font-bold">...</span>
            )}
        </p>
      ))}

      {content.length > sliceLength && (
        <div className="flex opacity-100">
          <button className="leading-none" onClick={toggleCollapse}>
            <small className="font-bold text-[13px]">
              {showAll ? "Show less" : "Show more"}
            </small>
          </button>
        </div>
      )}
    </li>
  );
};

ReviewItem.propTypes = {
  review: PropTypes.instanceOf(Object).isRequired,
};

export default ReviewItem;
