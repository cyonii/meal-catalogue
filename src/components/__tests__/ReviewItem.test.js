import { render, screen, cleanup } from '@testing-library/react';
import ReviewItem from '../ReviewItem';
import reviews from './data/reviews.json';

describe('ReviewItem', () => {
  afterEach(cleanup);

  it('should render correctly', () => {
    const review = reviews[0];
    render(<ReviewItem review={review} />);

    expect(screen.getByText(review.author)).toBeInTheDocument();
    expect(screen.getByText(`Rating: ${review.author_details.rating}`)).toBeInTheDocument();
  });

  it('should render `show more` button when content length > 250', () => {
    const review = reviews[0];
    render(<ReviewItem review={review} />);

    expect(screen.getByText('Show more')).toBeInTheDocument();
  });

  it('should NOT render `show more` button when content length < 250', () => {
    const review = reviews[1];
    render(<ReviewItem review={{ ...review, content: review.content.slice(0, 200) }} />);

    expect(screen.queryByText('Show more')).toBeNull();
  });

  it('should toggle between `show less` and `show more` button is clicked', () => {
    const review = reviews[0];
    render(<ReviewItem review={review} />);

    screen.getByText('Show more').click();
    expect(screen.queryAllByText('Show less')).toHaveLength(2);
    expect(screen.queryByText('Show more')).toBeNull();

    screen.queryAllByText('Show less')[0].click();
    expect(screen.getByText('Show more')).toBeInTheDocument();
    expect(screen.queryByText('Show less')).toBeNull();
  });
});
