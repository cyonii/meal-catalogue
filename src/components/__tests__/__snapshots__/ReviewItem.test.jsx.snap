// Vitest Snapshot v1

exports[`ReviewItem > should render correctly 1`] = `
<div>
  <li
    class="px-0 bg-transparent text-white border-0 border-bottom border-secondary list-group-item"
    data-testid="review-item"
  >
    <h6
      class="text-info mb-0 lh-1"
    >
      Cat Ellington
    </h6>
    <div
      class="d-flex justify-content-between align-items-center mb-3"
    >
      <small
        class="text-muted"
      >
        Rating: null
      </small>
    </div>
    <p
      class="mb-2 opacity-75 review-text"
    >
      (As I'm writing this review, Darth Vader's theme music begins to build in my mind...)
    </p>
    <p
      class="mb-2 opacity-75 review-text"
    >
      
    </p>
    <p
      class="mb-2 opacity-75 review-text"
    >
      Well, it actually has a title, what the Darth Vader theme. And that title is "The Imperial March", composed by the great John Williams, whom, as many of you may 
      <span
        class="text-muted fw-bold"
      >
        ...
      </span>
    </p>
    <div
      class="d-flex mt-1 opacity-100"
    >
      <button
        class="lh-1 py-0 pb-1 rounded-0 text-decoration-none btn btn-info"
        type="button"
      >
        <small
          class="fw-bold"
          style="font-size: 13px;"
        >
          Show more
        </small>
      </button>
    </div>
  </li>
</div>
`;
